src/metabase/api/timeline_event.clj

```diff
@@ -14,28 +15,37 @@
 
 (api/defendpoint POST "/"
   "Create a new [[TimelineEvent]]."
-  [:as {{:keys [name description timestamp time_matters timezone icon timeline_id archived] :as body} :body}]
+  [:as {{:keys [name description timestamp time_matters timezone icon timeline_id source question_id archived] :as body} :body}]
   {name         su/NonBlankString
    description  (s/maybe s/Str)
    timestamp    su/TemporalString
    time_matters (s/maybe s/Bool)
    timezone     s/Str
    icon         (s/maybe timeline-event/Icons)
    timeline_id  su/IntGreaterThanZero
+   source       (s/maybe timeline-event/Sources)
+   question_id  (s/maybe su/IntGreaterThanZero)
    archived     (s/maybe s/Bool)}
   ;; deliberately not using api/check-404 so we can have a useful error message.
   (let [timeline (Timeline timeline_id)]
     (when-not timeline
       (throw (ex-info (tru "Timeline with id {0} not found" timeline_id)
                       {:status-code 404})))
-    (collection/check-write-perms-for-collection (:collection_id timeline)))
-  ;; todo: revision system
-  (let [parsed (if (nil? timestamp)
-                 (throw (ex-info (tru "Timestamp cannot be null") {:status-code 400}))
-                 (u.date/parse timestamp))]
-    (db/insert! TimelineEvent (assoc body
-                                     :creator_id api/*current-user-id*
-                                     :timestamp parsed))))
+    (collection/check-write-perms-for-collection (:collection_id timeline))
+    ;; todo: revision system
+    (let [parsed (if (nil? timestamp)
+                   (throw (ex-info (tru "Timestamp cannot be null") {:status-code 400}))
+                   (u.date/parse timestamp))
+          tl-event (merge (dissoc body :source :question_id)
+                          {:creator_id api/*current-user-id*
+                           :timestamp parsed})]
+      (snowplow/track-event! ::snowplow/new-event-created
+                             api/*current-user-id*
+                             (cond-> {:time_matters time_matters
+                                      :collection_id (:collection_id timeline)}
+                               (boolean source) (assoc :source source)
+                               (boolean question_id) (assoc :question_id question_id)))
+      (db/insert! TimelineEvent tl-event))))
```

Comment:
I believe the db insert of `tl-event` here will be missing the `:source` and `:question_id` keys in this version, whereas previously we were inserting the body with those keys included. I'm not sure if those keys are needed! just pointing this out

Reply:
That's intentional.
:source and :question_id come from the frontend and are only used in the snowplow/track-event! The db does not have those columns so would fail if I don't dissoc those keys